src/core/net/ip6.cpp

```diff
@@ -1273,22 +1271,43 @@ otError Ip6::HandleDatagram(Message &aMessage, Netif *aNetif, const void *aLinkM
         hopLimit = header.GetHopLimit();
         aMessage.Write(Header::kHopLimitFieldOffset, hopLimit);
 
-#if OPENTHREAD_CONFIG_UNSECURE_TRAFFIC_MANAGED_BY_STACK_ENABLE
         // check whether source port is an unsecure port
+        if (aFromNcpHost && nextHeader == kProtoIcmp6)
+        {
+            Icmp::Header icmp;
+
+            SuccessOrExit(error = aMessage.Read(aMessage.GetOffset(), icmp));
+            VerifyOrExit(
+                icmp.GetType() != Icmp::Header::kTypeDstUnreach && icmp.GetType() != Icmp::Header::kTypePacketToBig &&
+                    icmp.GetType() != Icmp::Header::kTypeTimeExceeded &&
+                    icmp.GetType() != Icmp::Header::kTypeEchoRequest && icmp.GetType() != Icmp::Header::kTypeEchoReply,
+                error = OT_ERROR_DROP);
```

Comment:
As I see, `Icmp::Header::Type` can be:
```
        enum Type : uint8_t
        {
            kTypeDstUnreach       = OT_ICMP6_TYPE_DST_UNREACH,       ///< Destination Unreachable
            kTypePacketToBig      = OT_ICMP6_TYPE_PACKET_TO_BIG,     ///< Packet To Big
            kTypeTimeExceeded     = OT_ICMP6_TYPE_TIME_EXCEEDED,     ///< Time Exceeded
            kTypeParameterProblem = OT_ICMP6_TYPE_PARAMETER_PROBLEM, ///< Parameter Problem
            kTypeEchoRequest      = OT_ICMP6_TYPE_ECHO_REQUEST,      ///< Echo Request
            kTypeEchoReply        = OT_ICMP6_TYPE_ECHO_REPLY,        ///< Echo Reply
        };
```
If the type is `kTypeParameterProblem` or other undefined type, do we need to drop it?

Reply:
Do we really want to drop ICMPv6 Errors like Destination Unreachable?

Reply:
I think you actually want to accept those ICMP messages, right? @gjc13 

```suggestion
            VerifyOrExit(
                icmp.GetType() == Icmp::Header::kTypeDstUnreach || icmp.GetType() == Icmp::Header::kTypePacketToBig ||
                    icmp.GetType() == Icmp::Header::kTypeTimeExceeded ||
                    icmp.GetType() == Icmp::Header::kTypeEchoRequest || icmp.GetType() == Icmp::Header::kTypeEchoReply,
                error = OT_ERROR_DROP);
```

Reply:
Yes, wrong logic here.